### PR TITLE
Upgrade from `react-proxy@1/react-transform` to `react-hot-loader@3.x`

### PR DESCRIPTION
**Pull request by EugeneZ**
This is a rather large change. I've tried to update the docs as best I can to reflect the new changes, but I'm going to copy-paste the migration section from the README because I think it's a good summary of the changes:

Migration is easy:

1. Remove `react-proxy` and `babel-plugin-react-transform` dependencies from `package.json` using `npm uninstall` or `yarn remove`
2. Use `npm install` or `yarn add` to get the latest `livereactload` and `react-hot-loader` packages.
3. Replace the `react-transform` section of your Babel config with `react-hot-loader/babel`. See installation instructions for how the config should look now. (It's must simpler.) 
4. Add `import { AppContainer} from 'react-hot-loader'` to the file where you render your root component, and then wrap your root component in this container. `ReactDOM.render(<App/>, ...)` becomes
`ReactDOM.render(<AppContainer><App></AppContainer>, ...)`.
5. If you were using `module.onReload` it must be replaced with `module.hot.accept`. This isn't simply a rename, this hook works differently now. Despite this, if you were following the commonly used patterns for this hook in the LiveReactload examples, simply renaming it should work just fine.

**You're all set!** Despite the easy migration, this is a rather large change under the hood so while reloading is improved in general, you may run into other issues. The biggest change to LiveReactload is that in the past, if only React components were updated, your application would not be re-run, and `react-proxy` would silently replace components. This was a bad approach as it circumvents the standard React approach of re-rendering to get changes. So now whenever you change a React component (or any file) your whole app will be re-run by default. If you have functionality that cannot be run multiple times on the same page load, you may need to use a `module.hot.accept` hook to have more control over this. See examples above!

### CODE REVIEW

**Comment by EugeneZ** [comment]
I think the Travis build just ran out of time or something, works fine locally.

**Comment by milankinen** [comment]
Yes, it's some kind of timing bug in tests. Fix'd

File: README.md [hunk 1]
```diff
@@ -50,73 +58,102 @@ If you want to stick with browserify, but use the Hot Module Reloading API (like
...
+)
+```
+
 **That's it!** Now just start (live) coding! For more detailed example, please see
 **[the basic usage example](examples/01-basic-usage)**.
 
+**NOTE:** The above snippet is idempotent, which means it can be run over and over without issue. Sometimes
+you may place code here (like `redux` or `react-tap-event-plugin`) which are stateful and cannot be run over
+and over. In this case, you must use `module.hot.onUpdate` to indicate to LiveReactload what to do when a change
+is detected. See the []non-idempotent example](examples/04-non-idempotent).
```
Comment (milankinen) [inline README.md]: Extra `]`

File: README.md [hunk 2]
```diff
@@ -50,73 +58,102 @@ If you want to stick with browserify, but use the Hot Module Reloading API (like
...
+import getStore from './getStore'
+
+ReactDOM.render(
+  <AppContainer>
+    <MyComponent />
+  </AppContainer>,
+  document.getElementById('app')
+)
+
+// getStore.js
```
Comment (milankinen) [inline README.md]: Can we split these to two separate code blocks? 

In addition, I'm not sure what is the meaning of this code example because `module.hot.onUpdate` is not used even though it's mentioned above? If the purpose is to demonstrate non-idempotent example, could that be mentioned before the snipped?

File: README.md [hunk 3]
```diff
@@ -50,73 +58,102 @@ If you want to stick with browserify, but use the Hot Module Reloading API (like
...
+Use LiveReactload as a Browserify plugin with `watchify`. For example:
 
 ```bash
 node_modules/.bin/watchify site.js -t babelify -p livereactload -o static/bundle.js
 ```
 
+Finally, when you create your application, the file where you render the top of the tree
+should look something like this:
+
+```javascript
```
Comment (milankinen) [inline README.md]: Do we need to import `patch` here?

Comment (milankinen) [inline README.md]: Nevermind. Should have read the instructions more carefully.

File: src/reloading.js [hunk 4]
```diff
@@ -123,12 +114,20 @@ function loader(mappings, entryPoints, options) {
...
-        __accepted: false,
-        onReload: function (hook) {
-          scope.reloadHooks[id] = hook;
+        hot: {
+          onUpdate: function (maybe, hook) {
+            var realHook = hook;
+            if (!realHook) {
+              realHook = maybe;
+            } else {
+              console.warn(`LiveReactload: You are providing two arguments to the module.hot.onUpdate hook, and we are
```
Comment (milankinen) [inline src/reloading.js]: Now that we're using `onUpdate` instead of `accept`, perhaps we could use `console.error`?

File: test/app/extra.js [hunk 5]
```diff
@@ -1,15 +1,27 @@
 import React from "react"
-import Header from "./.src/extra/header"
-import Body from "./.src/extra/body"
+
+import Root from "./.src/extra/root"
 
 class Box extends React.Component {
+  constructor(props) {
```
Comment (milankinen) [inline test/app/extra.js]: Hmmm this looks a bit nasty but have to figure out better way later... 🤔

File: test/smokeTest.js [hunk 6]
```diff
@@ -147,13 +147,13 @@ test("smoke tests", assert => {
         ])
       return updateSrcP
         .then(() => {
-          assert.equals(browser.window._noAcceptReloaded, true)
+          assert.equals(browser.window._noAcceptReloaded, false)
```
Comment (milankinen) [inline test/smokeTest.js]: Why wasn't this variable changed? It should call `onUpdate` though?

File: test/app/site.js [hunk 7]
```diff
@@ -1,6 +1,7 @@
+import patch from "react-hot-loader/patch"
```
Comment (milankinen) [inline test/app/site.js]: Why is this needed here?

File: README.md (hunk 2 above)
Comment (EugeneZ) [inline README.md]: Yeah, I changed the README several times, I think the purpose of this example got lost at some point... I'll improve it

File: src/reloading.js (hunk 4 above)
Comment (EugeneZ) [inline src/reloading.js]: Well, I think `console.error` indicates that something broke... since we're no longer using `accept` maybe we shouldn't accept two arguments at all though.

File: test/app/extra.js (hunk 5 above)
Comment (EugeneZ) [inline test/app/extra.js]: Yeah, this is a bit of a problem. Because with `react-proxy@1` it would "detect" the change and force the update automatically, but `react-hot-loader@3` doesn't do that anymore (for good reason), but it means stuff like external components are much harder to live update because something has to load the new module and force the re-render. Kind of like what `AppContainer` does. In fact maybe we can just wrap this component in `AppContainer` and that would fix the issue. Agree this isn't perfect but I figured it was okay for a test.

File: test/smokeTest.js (hunk 6 above)
Comment (EugeneZ) [inline test/smokeTest.js]: Are you saying the naming is confusing now? I agree, but not sure what a better name is. I tried changing the locations of `true` and `false` and I think it makes more sense now.

**Comment by EugeneZ** [comment]
Fixed most of your PR comments except the weird test and and the warning from the hook.

File: src/reloading.js (hunk 4 above)
Comment (milankinen) [inline src/reloading.js]: Yeah. Let's drop the second argument entirely 👍

**Comment by milankinen** [comment]
Looks good! 👍 There is still one tiny issue in `test/app/src/hooks/noAccept.js` (assignment should be `window._noAcceptReloaded = window._noAcceptReloaded || false`) but I can fix it later. I'm gonna merge this and make the first beta release tonight.

**Comment by milankinen** [comment]
Okay! Sorry this took a bit longer than promised but the changes are now merged to `4.x` branch and the new version is relesed under `livereactload@next` tag.

I made some minor changes to hooks test code and fixed the version numbers from examples. Link to the new version is also added to the top of README in `master` branch. Let's hope we get some people to test this new version! 👍

**Comment by fibo** [comment]
Hi all, I tested it and it almost worked. I do not know if I missed something, any hint?

Please follow these [instructions to reproduce it](https://github.com/fibo/zeroconf-redux#livereactload). I used a copy of the classic counter example, [you can find here](https://github.com/fibo/zeroconf-redux/tree/master/examples/counter).

This is the result I get, the state was not erased but I got also brand new DOM nodes added.

screen shot 2017-05-05 at 21 14 26

**Comment by EugeneZ** [comment]
@fibo 

Hello! The problem with your approach is here:

    const root = document.createElement(root)
    document.body.appendChild(root)

This code is not idempotent, meaning that it can safely be run multiple times. When run multiple times, it creates and appends a node each time, hence the issue you are describing.

To fix this, you can take a number of approaches:

1. Recommended: Make your code idempotent by giving the created node an id, and checking for that idea before creating and appending the node.
2. Make your code idempotent by not creating the root element in code, instead use an HTML file with a prexisiting node to mount on
3. If you don't want to change your code like that (understandable) then you need to tell LiveReactload how to handle your non-idempotent code. Put this anywhere in the file: `if (module && module.hot) { module.hot.onUpdate(()=>render(<AppContainer><Component /></AppContainer>,root))}`

As you can see, the point is that without this hook, the whole file is re-run when changed. When you have the hook, we need to handle the update manually, and we need to re-render the React tree.

Does that make sense?

**Comment by fibo** [comment]
Thank you very much, it works!
